consensus/src/block_storage/execution_pool/target_block_retrieval_test.rs

```diff
@@ -0,0 +1,295 @@
+    let highest_quorum_cert = a4_r9.quorum_cert().clone();
+    let highest_quorum_cert_round = highest_quorum_cert.certified_block().round();
+    let highest_quorum_cert_id = highest_quorum_cert.certified_block().id();
+    let highest_commit_cert = highest_quorum_cert.into_wrapped_ledger_info();
+    let highest_commit_cert_round = highest_commit_cert.commit_info().round();
+
+    // commit_root, window_root (my validator)       highest_quorum_cert (different validator)
+    //      ┌───────────────┘             ┌─────────────────────────────┘
+    //      ↓                             ↓
+    //  Genesis ──> A1_R1 ──> A2_R3 ──> A3_R6 ──> A4_R9
+    //      ↑
+    //      └────────────┐
+    //    highest_commit_cert (different validator)
+    assert_eq!(commit_root, genesis_block.id());
+    assert_eq!(window_root, genesis_block.id());
+    assert_eq!(highest_quorum_cert_round, 6);
+    assert_eq!(highest_commit_cert_round, 0);
+
+    let (payload, num_blocks) = BlockStore::generate_target_block_retrieval_payload_and_num_blocks(
+        &highest_quorum_cert,
+        &highest_commit_cert,
+        window_size,
+    );
+
+    match payload {
+        TargetBlockRetrieval::TargetBlockId(id) => {
+            assert_eq!(id, genesis_block.id());
+            assert_eq!(num_blocks, 7);
+
+            let request =
+                BlockRetrievalRequest::V1(BlockRetrievalRequestV1::new_with_target_block_id(
+                    highest_quorum_cert_id,
+                    num_blocks,
+                    id,
+                ));
+            let response = block_store.process_block_retrieval_inner(&request).await;
+            let blocks = response.blocks();
+
+            assert_eq!(
+                blocks.first().expect("No first block found").round(),
+                _a3_r6.block().round()
+            );
+            assert_eq!(
+                blocks.last().expect("No last block found").round(),
+                genesis_block.block().round()
+            );
+
+            // Verifies BlockRetrievalStatus and num_blocks and target_block_id matches
+            assert_ok!(response.verify_inner(&request));
+        },
+        TargetBlockRetrieval::TargetRound(_) => {
+            panic!("Should not be TargetRound variant")
+        },
+    }
+}
+
+#[tokio::test]
+async fn test_window_quorum_round_greater_than_commit_round() {
+    let window_size: Option<u64> = Some(1u64);
+    let (_, block_store, pipelined_blocks) =
+        create_block_tree_with_forks_unordered_parents(window_size).await;
+    let [_genesis_block, a1_r1, a2_r3, a3_r6, a4_r9, _b1_r2, _b2_r4, _b3_r5, _c1_r7, _d1_r8] =
+        pipelined_blocks;
+
+    // Prune tree, moves the commit root to a2_r3 and move the highest_commit_root too
+    block_store.commit_callback(
+        a2_r3.block().id(),
+        a2_r3.block().round(),
+        a3_r6.quorum_cert().into_wrapped_ledger_info(), // TODO this is correct right?
```

Comment:
technically this is not right, quorum cert can only commits the previous round if it's continuous i.e. QC for R6 can only commit if the previous round is R5. but not sure if it matters here

Reply:
hmm is there another way to get `wrapped_ledger_info` for `a2_r3` @zekun000?

Reply:
there should be/we could have a function to create a ledger info that points to a2_r3 and turns it to the wrapped version

Reply:
I think I tried to do this already here, which is why it works in the test even though the round is not consecutive: https://github.com/aptos-labs/aptos-core/blob/12e6290a02a6a842f998d90af33ccf209258fa30/consensus/src/block_storage/execution_pool/common_test.rs#L232-L232. 

Normally the third parameter for `insert_block(...)` is None unless you want to specify the block info that's used for the QC.